azure-iot-device/azure/iot/device/common/pipeline/pipeline_stages_base.py

```diff
@@ -581,6 +585,26 @@ def _handle_pipeline_event(self, event):
                         self.name, event.name, event.request_id
                     )
                 )
+
+        elif isinstance(event, pipeline_events_base.ConnectedEvent):
+            """
+            If we're reconnecting, send all pending requests down again.  This is necessary
+            because any response that might have been sent by the service was likely lost
+            when the connection dropped.  The fact that the operation is still pending means
+            that we haven't received the response yet, and it's safe to re-send the request
+            because everything we do with this stage is allowed to repeat.  For instance, if
+            we update a twin twice with the same reported properties, the service will not
+            fail, likewise for getting a twin, or registering with the provisioning service.
```

Comment:
Will it add spurious increments to the $version though?

Reply:
Yes, but that's a problem for a different layer.  At this layer, our goal is to ensure delivery.  We can do that, and we can verify it with tests.  Beyond that is uncharted territory.  I'm not sure if there's even a well-defined contract for the behavior of the $version attribute.  I know that it goes up by sometimes, but what happens if the server received a PATCH with an already used $rid?  Surely it _could_ know that this is a duplicate message and behave accordingly, but is it guaranteed to?  What about PATCH operations that don't actually change any twin properties?  Do those update $version even though nothing is actually changing?  I'm pretty sure that they do, but is this by coincidence or by design?  This is a rat's nest and we can't solve it here. 

---
In reply to: [400433963](https://github.com/Azure/azure-iot-sdk-python/pull/520#discussion_r400433963) [](ancestors = 400433963)